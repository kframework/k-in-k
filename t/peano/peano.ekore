module A-BASIC-K
  sort K{} []
  sort KItem{} []
endmodule [ ]

module B-KSEQ
  import A-BASIC-K []
  symbol kseq{}(KItem{},K{}) : K{} []
  symbol append{}(K{}, K{}) : K{} [function{}()]
  symbol dotk{}() : K{} []
  axiom { R } \equals { K { }, R } ( append { } ( dotk { } ( ) , K2:K { } ) , K2 : K { } ) [ ]
  axiom { R } \equals { K { }, R } ( append { } ( kseq { } ( K1 : KItem { } , K2 : K { } ) , K3 : K { } ) , kseq { } ( K1 : KItem { } , append { } (K2 : K { } , K3 : K { } ) ) ) [ ]
endmodule [ ]

module C-INJ
  symbol inj{From,To}(From) : To []
  axiom{S1,S2,S3,R} \equals{S3,R}( inj{S2,S3}(inj{S1,S2}(T:S1)), inj{S1,S3}(T:S1)) []
endmodule [ ]

module D-K
  import B-KSEQ []
  import C-INJ []
endmodule [ ]

module PEANO
  import D-K []

  syntax Nat ::=
      "zero" [ klabel ( zero ), constructor, functional ]
    | "succ" "(" Nat ")" [ klabel ( succ ), constructor, functional ]
    | "plus" "(" Nat "," Nat ")" [ klabel ( plus ), function ]
  rule plus{}(zero{}(), Y:Nat{}) => Y:Nat{}
  rule plus{}(succ{}(X:Nat{}), Y:Nat{}) => succ{}(plus{}(X:Nat{}, Y:Nat{}))
endmodule
