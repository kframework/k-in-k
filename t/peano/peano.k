module PEANO
  syntax Pgm ::= Nat
  syntax Layout ::= r"(/\\*([^\\*]|(\\*+([^\\*/])))*\\*+/|//[^\n\r]*|[\\ \n\r\t])*" [klabel(layout)]

  syntax Nat ::= "zero" [klabel(zero)]
               | succ(Nat) [klabel(succ)]
               | plus(Nat, Nat) [function, klabel(plus)]

  rule plus(zero, Y:Nat) => Y
  rule plus(succ(X:Nat), Y:Nat) => succ(plus(X, Y))
endmodule
