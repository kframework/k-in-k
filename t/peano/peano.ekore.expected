[  ]

module PEANO
  symbol zero { } (  ) : Nat { } [ ]
  symbol succ { } ( Nat { } ) : Nat { } [ ]
  symbol plus { } ( Nat { } , Nat { } ) : Nat { } [ function { } ( ) ]
  sort Nat { } [ ]
  syntax Nat ::=
    "zero" [ klabel ( zero ) ]
    | "succ" "(" Nat ")" [ klabel ( succ ) ]
    | "plus" "(" Nat "," Nat ")" [ klabel ( plus ), function ]
  rule plus(zero(), Y:Nat) => Y
  rule plus(succ(X:Nat), Y:Nat) => succ(plus(X, Y))

endmodule [  ]

